### PR TITLE
feat: drop Python 3.9 and test on Python 3.10/3.12

### PR DESCRIPTION
**Pull request by deepyaman**
## Description of changes

According to NEP-29, we could have dropped Python 3.9 last month.

* Test on 3.10/3.12 in CI, except for a few exceptions (e.g. Flink).
* Apply `pyupgrade` rules to remove code for Python<3.10.
* Remove workaround for DuckDB geospatial tests.

## Issues closed

N/A

### CODE REVIEW

File: ci/make_geography_db.py
Comment (deepyaman) [inline ci/make_geography_db.py:66]: Kind of annoying that this happened, but the reason is that https://docs.astral.sh/ruff/rules/zip-without-explicit-strict/ is available when targeting 3.10.

Guess these could mostly be `strict=True`, or could just ignore B905 for stylistic reasons.

Comment (jcrist) [reply ci/make_geography_db.py:66]: I'd vote to just disable that rule.

File: ibis/backends/bigquery/__init__.py
Comment (deepyaman) [reply ibis/backends/bigquery/__init__.py:1034]: Yeah, for sure. Turns out it's a separate rule: https://docs.astral.sh/ruff/rules/non-pep604-isinstance/

And the docs even say it's slower. I'm surprised they'd enable that by default...

Comment (deepyaman) [reply ibis/backends/bigquery/__init__.py]: Done

File: ci/make_geography_db.py
Comment (deepyaman) [reply ci/make_geography_db.py]: and done!

**Review by jcrist — APPROVED** [review]
Provided tests pass, this looks good to me! We should get a review by one other person before merging (@gforsyth?) just to ensure I didn't miss anything around the CI setup.

**Comment by deepyaman** [comment]
> Provided tests pass, this looks good to me! We should get a review by one other person before merging (@gforsyth?) just to ensure I didn't miss anything around the CI setup.

FYI tests have passed; the nix 3.9 test just won't complete, because it's removed.

**Comment by cpcloud** [comment]
I'll remove the required nix 3.9 test

**Comment by cpcloud** [comment]
@jcrist @deepyaman I think historically we made this a breaking change. Should we considering changing that convention?

File: .github/workflows/ibis-backends.yml
Comment (ncclementi) [inline .github/workflows/ibis-backends.yml:240]: shouldn't this be the 3.10, if it was 3.9 before?

**Comment by jcrist** [comment]
> I think historically we made this a breaking change. Should we considering changing that convention?

I think that dropping a Python version in a minor release seems fine (it's also pretty common in the ecosystem). Since we've updated the minimum version requirement, even pip's solver won't allow them to install the new release with an older version of python. The "breakage" here isn't something a user should be able to encounter.

File: .github/workflows/ibis-backends.yml
Comment (deepyaman) [reply .github/workflows/ibis-backends.yml:240]: The diff is a bit weird. It was a Dask entry before, this one is a Flink entry. The reason we test on 3.11 for Flink still is that PyFlink doesn't support 3.12.

**Review by cpcloud — APPROVED** [review]
LGTM.

Can we make this a feature or fix PR?

**Comment by deepyaman** [comment]
> LGTM.
> 
> Can we make this a feature or fix PR?

Done! Also rebased.

**Pull request merged**
